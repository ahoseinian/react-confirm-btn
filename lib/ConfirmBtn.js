// @flow

import React from 'react';
import CountDown from './CountDown';

type props = {
  children?: any,
  confirmText?: string,
  time?: number,
  style?: {},
  onConfirm: () => void,
};

class ConfirmButton extends React.Component {
  props: props;
  state = { confirmed: false, showCountDown: false };
  static defaultProps = {
    confirmText: 'Sure?',
  };

  checkConfirmation = () => {
    const { confirmed } = this.state;
    const { time } = this.props;

    if (confirmed) {
      this.props.onConfirm();
      this.setState({ showCountDown: false });
    } else if (time) {
      this.setState({ showCountDown: true });

      setTimeout(
        () => this.setState({ confirmed: false, showCountDown: false }),
        time,
      );
    }

    this.setState({ confirmed: !confirmed });
  };

  render() {
    const { confirmed, showCountDown } = this.state;
    const {
      children,
      confirmText,
      style,
      time,
      onConfirm,
      ...rest
    } = this.props;

    return (
      <button
        {...rest}
        style={{ ...s.button, ...style }}
        onClick={this.checkConfirmation}
      >
        {confirmed ? <span>{confirmText}</span> : children}
        {showCountDown && <CountDown time={time} />}
      </button>
    );
  }
}

export default ConfirmButton;

const s = {
  button: {
    position: 'relative',
  },
};
