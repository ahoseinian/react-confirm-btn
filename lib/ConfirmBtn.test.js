// @flow

import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ConfirmBtn from './ConfirmBtn';

const getProps = () => ({
  onConfirm: jest.fn(),
});
const wait = (ms = 100) => new Promise(r => setTimeout(() => r(true), ms));

it('renders', () => {
  const wrapper = shallow(<ConfirmBtn {...getProps()}> Delete </ConfirmBtn>);

  expect(toJson(wrapper)).toMatchSnapshot();
});

it('passes the props down to the button', () => {
  const style = { background: 'black' };
  const wrapper = shallow(
    <ConfirmBtn {...getProps()} style={style}> Delete </ConfirmBtn>,
  );

  expect(wrapper.prop('style').background).toBe(style.background);
});

it('changes confirmed state when clicked', () => {
  const wrapper = shallow(<ConfirmBtn {...getProps()}> Delete </ConfirmBtn>);
  wrapper.simulate('click');

  expect(wrapper.state().confirmed).toBe(true);
});

it('changes confirmed to false after doing onConfirm()', () => {
  const props = getProps();
  const wrapper = shallow(<ConfirmBtn {...props}> Delete </ConfirmBtn>);
  wrapper.setState({ confirmed: true });
  wrapper.simulate('click');

  expect(props.onConfirm).toBeCalled();
  expect(wrapper.state().confirmed).toBe(false);
});

it('shows CountDown when time is set', () => {
  const wrapper = shallow(
    <ConfirmBtn {...getProps()} time={2000}> Delete </ConfirmBtn>,
  );
  wrapper.simulate('click');

  expect(wrapper.find('CountDown').length).toBe(1);
});

it('returns the state to unconfirmed after timeout', async () => {
  const wrapper = shallow(
    <ConfirmBtn {...getProps()} time={100}> Delete </ConfirmBtn>,
  );
  wrapper.simulate('click');

  expect(wrapper.state()).toEqual({ confirmed: true, showCountDown: true });
  await wait(100);
  expect(wrapper.state()).toEqual({ confirmed: false, showCountDown: false });
});
