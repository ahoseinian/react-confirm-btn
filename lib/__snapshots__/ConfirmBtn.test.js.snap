// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<button
  onClick={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
>
   Delete 
</button>
`;
